ext/standard/string.c

```diff
@@ -4790,11 +4790,29 @@ PHP_FUNCTION(strip_tags)
 		Z_PARAM_ZVAL(allow)
 	ZEND_PARSE_PARAMETERS_END();
 
-	/* To maintain a certain BC, we allow anything for the second parameter and return original string */
 	if (allow) {
-		convert_to_string(allow);
-		allowed_tags = Z_STRVAL_P(allow);
-		allowed_tags_len = Z_STRLEN_P(allow);
+		if (Z_TYPE_P(allow) == IS_ARRAY) {
+			smart_str tags_ss = {0};
+			zval *tmp;
+
+			smart_str_alloc(&tags_ss, 0, 0);
+			ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(allow), tmp) {
+				convert_to_string_ex(tmp);
+				smart_str_appendc(&tags_ss, '<');
+				smart_str_append(&tags_ss, Z_STR_P(tmp));
+				smart_str_appendc(&tags_ss, '>');
+			} ZEND_HASH_FOREACH_END();
+			smart_str_0(&tags_ss);
+			allowed_tags = ZSTR_VAL(tags_ss.s);
+			allowed_tags_len = ZSTR_LEN(tags_ss.s);
+			smart_str_free(&tags_ss);
```

Comment:
This is going to free the allowed_tags string, even though it will be used later. The free call for the smart_str should occur only after the php_strip_tags_ex call.